src/features/Cart/hooks/useCart.ts

```diff
@@ -0,0 +1,88 @@
+import { useContext, useEffect, useRef, useState } from 'react';
+
+import { deleteCartItem, getCartItemList, updateCartItem } from '@/api/cart';
+import { ToastContext } from '@/shared/context/ToastProvider';
+import { useFetchData } from '@/shared/hooks/useFetchData';
+import { isError } from '@/shared/utils/isError';
+
+import { CartItem } from '../types/Cart.types';
+
+export const useCart = () => {
+  const cart = useFetchData<CartItem[]>({ autoFetch: getCartItemList });
+  const { showToast } = useContext(ToastContext);
+  const [checkedItems, setCheckedItems] = useState<Set<number>>(new Set());
+  const hasInitialized = useRef(false);
+
+  useEffect(() => {
+    if (cart.data && cart.data.length > 0 && !hasInitialized.current) {
+      setCheckedItems(new Set(cart.data.map((item) => item.id)));
+      hasInitialized.current = true;
+    }
+  }, [cart.data, checkedItems.size]);
```

Comment:
이와 같은 초기화 로직은 사실 cart 데이터를 초기 마운트 시점에 받을 수만 있다면 useState 만으로 처리할 수 있다고 생각합니다. 그렇게 되면 hasInitialized 와 같은 ref 도 사용할 필요가 없어집니다.
```suggestion
export const useCart = (cart: CartItem[]) => {
  ...
  const [checkedItems, setCheckedItems] = useState<Set<number>>(new Set(cart.data.map((item) => item.id));
```

Reply:
`useFetchData`는 비동기 훅인데요. 컴포넌트가 마운트 되는 시점에는 `cart.data`는 항상 `undefined`가 나오게 됩니다. (아직 useEffect를 통해 동기화 하는과정이 없기 때문에요!) 따라서 `초기 마운트 시점에 받을 수만 있다면` 이라고 말씀해주신 부분이 이해가 잘 가지 않습니다!

또 `useState`의 초기값은 컴포넌트가 처음 마운트될 때만 설정되고, 이후 cart.data가 로드되어도 `checkedItems`는 업데이트되지 않을 것 같은데요. (아까 공식문서 Props를 state에 미러링하지 마세요에 나와 있는 설명처럼요!) 이에 대해서도 어떻게 생각하시는지 궁금합니다~!

Reply:
cart 가 아니라 cart.data 를 명시했었어야 했는데 실수가 있었던 것 같네요. 제가 말씀드린 `cart 데이터를 초기 마운트 시점에 받을 수만 있다면 ` 라는 말은 useFetchData 훅으로 받게 되는 cart 상태가 undefined 가 아닌 시점에 주요 로직이 잠긴 컴포넌트를 렌더링 하는 방식을 뜻한 것이었습니다. 예시로 든 아래 코드를 생각해볼 수 있을 것 같네요.

```tsx
const CartPageFunnel = ({ initialCartItems }: { initialCartItems: CartItem[] }) => {
  const { Funnel, setStep } = useFunnel<STEPS>('장바구니');
  const [cartItems, setCartItems] = useState<CartItem[]>(initialCartItems);
  return (
    <Funnel>
      <Funnel.Step name="장바구니">
        <CartInfo cartItems={cartItems} onCartItemsChange={setCartItems} />
        <OrderConfirmNavigationButton onClick={() => setStep('주문정보')} />
      </Funnel.Step>
      <Funnel.Step name="주문정보">
        <OrderConfirm cartItems={cartItems} onPrev={() => setStep('장바구니')} />
      </Funnel.Step>
    </Funnel>
  );
};

export const CartPage = () => {
  const { data: cart, isLoading } = useFetchData<CartItem[]>({ autoFetch: getCartItemList });
  if (isLoading) {
    return <div>Loading...</div>;
  }
  if (!cart || cart.data.length === 0) {
    return <div>장바구니가 비어있습니다.</div>;
  }
  return <CartPageFunnel initialCartItems={cart.data} />;
};
```

위의 예시에서, CartPage 는 cart.data 가 null 이 아닐 때 비로소 CartPageFunnel 컴포넌트를 렌더링하게 됩니다. 이렇게 처리하면 CartPageFunnel 은 반드시 initialCartItems 배열을 props 으로 받은 상태에서 초기 마운트를 하게 되므로 useState 을 사용한 cartItems 상태의 초기값은 initialCartItems 배열과 동일해집니다.

이렇게 처리하는 것과 아래와 같이 처리하는 것이 결과적으로는 동일한 결과를 만들어내기 때문에 이를 제안드린 것이었습니다.

```tsx
  useEffect(() => {
    if (cart.data && cart.data.length > 0 && !hasInitialized.current) {
      setCheckedItems(new Set(cart.data.map((item) => item.id)));
      hasInitialized.current = true;
    }
  }, [cart.data, checkedItems.size]);
```

Reply:
세라가 말씀주신 '미러링'을 지양하는 관점과 이전 미션부터 중요시 여겨졌던 SSOT 원칙을 고려해봤을 때, useFetchData 훅을 기반으로 받아온 cart 상태를 바탕으로 새로운 cartItems 상태를 초기화하는 것보다 useFetchData 훅이 cart 상태를 업데이트 할 수 있는 함수를 반환하도록 만드는 것이 더 나은 선택이라고 생각이 듭니다.

cart 상태와 cartItems 상태가 별도의 useState 훅을 기반으로 생성되는 경우, 두 상태를 서로 동기화 하는 등의 추가적인 로직이 필요해지며 동기화 로직이 지금처럼 한번만 실행되는 구조에서는 cart 와 cartItems 상태가 서로 일치하지 않을 위험도 생기기 때문입니다.

세라는 이에 대해서는 어떻게 생각하실까요?

Reply:
예시 써주셔서 완전 이해했어요! `cart.초기로딩`이 끝나고 난 뒤에 처리해주면 useEffect를 사용하지 않아도 되겠네요! 요 부분 `isInitialLoading` 상태 추가해서 수정했습니닷. 
[refactor: useFetchData hook에 isInitialLoading 추가](https://github.com/woowacourse/react-shopping-cart/pull/326/commits/aa21cf3d4588c2399855225555dba9def6bade89)
[refactor: CartPage 컴포넌트 조건문을 통한 분기 처리](https://github.com/woowacourse/react-shopping-cart/pull/326/commits/754b67a88a6a2a327c7b45617231b45701455350)

> 세라가 말씀주신 '미러링'을 지양하는 관점과 이전 미션부터 중요시 여겨졌던 SSOT 원칙을 고려해봤을 때, useFetchData 훅을 기반으로 받아온 cart 상태를 바탕으로 새로운 cartItems 상태를 초기화하는 것보다 useFetchData 훅이 cart 상태를 업데이트 할 수 있는 함수를 반환하도록 만드는 것이 더 나은 선택이라고 생각이 듭니다.

저도 이부분에 대해서는 동의합니다..! `useFetchData` 내부에서 setData를 업데이트 해주는 함수를 명시해 해당 문제를 해결할 수도 있지만 해당 훅이 서버 데이터 fetching이라는 명확한 책임을 가지고 있는데, 여기에 클라이언트 상태 관리까지 추가하면 훅의 책임이 과도하게 커지지 않을까 고민이 되더라구요! 

또 서버 상태를 관리하고 있다보니 setter를 반환해주는게 맞나..?하는 고민이 계속 드는 것 같아요! 두 상태가 존재할 때 휴먼에러로 인한 isChecked 값 불일치 문제도 어느 정도 고려하고 있습니다. 

조금 더 고민해보고 2단계에서 해당 주제에 대해 다시 얘기해봐도 괜찮을까요?!

Reply:
넵 물론입니다! 2단계에서는 고민하고 계신 문제에 대한 답이 될 수도 있는 'optimistic updates' 에 대해서도 생각해보시면 좋을 것 같습니다!

https://medium.com/@kyledeguzmanx/what-are-optimistic-updates-483662c3e171